pgwire: move the pre-serve logic into a separate file

knz opened the pull request:
Parent PRs:
- [x] https://github.com/cockroachdb/cockroach/pull/84608
- [x] https://github.com/cockroachdb/cockroach/pull/91739
- [x] https://github.com/cockroachdb/cockroach/pull/91744
- [x] https://github.com/cockroachdb/cockroach/pull/92574
- [x] https://github.com/cockroachdb/cockroach/pull/92575
- [x] https://github.com/cockroachdb/cockroach/pull/92576


This PR moves codes into the new file pre_serve.go, where the tenant-independent logic will live.

Epic: CRDB-14537

knz committed:
pgwire: unexport ServerMetrics, rename to tenantSpecificMetrics

Release note: None

knz committed:
pgwire: fix typo in metric descriptions

Release note: None

knz committed:
pgwire: create a framework for pre-serve connection logic

This commit introduces `PreServeConnHandler`, which will
handle the preparation of a SQL connection prior to handing it off to
a specific tenant server.

Release note (ops changes): The count of network bytes sent
to report re-authentication errors to a SQL client is now
reported via the metric `sql.pre_serve.bytesout` (instead of
`sql.bytesout` previously). The count of pre-authentication errors
is now reported via the metric `sql.pre_serve.conn.failures` (instead
of `sql.conn.failures` previously).

Release note (ops changes): The count of new SQL connections
is now also reported on `sql.pre_serve.new_conns`.

knz committed:
pgwire: simplify writeErr/sendErr

Release note: None

knz committed:
pgwire: lift readVersion + TLS handshake into the pre-serve handler

Release note (ops change): The bytes read from SQL clients
prior to authentication are now reported via the metric
`sql.pre_serve.bytesin` (instead of `sql.bytesin` previously).

knz committed:
pgwire: make handleCancel unaware of conn/buf

Release note: None

knz committed:
pgwire: split `parseClientProvidedSessionParameters`

We need the read/parse of the parameters to be tenant-independent,
so the function cannot contain tenant-specific configuration.

Release note: None

knz committed:
pgwire: lift the pre-serve logic into a closure

This demonstrates that the logic is tenant-independent
and makes it ready to move into the PreServe() function.

Release note: None

knz committed:
pgwire: split the pre-conn reserved memory account

This commit extracts the memory reservation into
a tenant-independent memory pool.

Release note: None

knz committed:
pgwire: move PreServe() to its own method

Release note: None

knz committed:
pgwire: move  maybeUpgradeToSecureConn to pre_serve.go

Release note: None

knz committed:
pgwire: move readVersion to pre_serve.go

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92577)
<!-- Reviewable:end -->

knz committed:
pgwire: move the parsing logic for client opts to a new file

Release note: None

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92577#-:-NI3dr6b7QOJtoy9buuW:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz committed:
cli/demo: fix the port allocation

Prior to this patch, we were unconditionally applying a port offset
for secondary tenant servers, even when the port allocation for the KV
node was randomized. This resulted in the demo cluster trying to
reuse arbitrary TCP ports which may be already in use elsewhere in the
system.

This patch fixes it by ensuring that tenant servers also use random
allocation when the KV node ports were randomly allocated in the first
place.

It also adds demo networking tests that verify this via the
`--sql-port` / `--http-port` flags.

Release note: None

knz committed:
cli/demo: deflake some interactive tests

Some of the tests were not waiting for the prompt to start sending
their input. This was causing the input to be lost in some cases.

Release note: None
